runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+		writeThread, err = strconv.Atoi(write.Threads)
+		if err != nil {
+			return fmt.Errorf("unable to parse write threads: %w", err)
+		}
+	}
+
+	if readThread == 0 && writeThread == 0 {
+		connector, err := duckdb.NewConnector("", nil)
+		if err != nil {
+			return fmt.Errorf("unable to create duckdb connector: %w", err)
+		}
+		defer connector.Close()
+		db := sql.OpenDB(connector)
+		defer db.Close()
+
+		row := db.QueryRow("SELECT value FROM duckdb_settings() WHERE name = 'threads'")
+		var threads int
+		err = row.Scan(&threads)
+		if err != nil {
+			return fmt.Errorf("unable to get threads: %w", err)
+		}
+
+		read.Threads = strconv.Itoa((threads + 1) / 2)
+		write.Threads = strconv.Itoa(threads / 2)
+	}
+
+	if readThread == 0 != (writeThread == 0) {
+		// only one is defined
+		var threads int
+		if readThread != 0 {
+			threads = readThread
+		} else {
+			threads = writeThread
+		}
+
+		read.Threads = strconv.Itoa((threads + 1) / 2)
+		write.Threads = strconv.Itoa(threads / 2)
+	}
+
+	err = mapstructure.WeakDecode(read, &d.ReadSettings)
+	if err != nil {
+		return fmt.Errorf("failed to update read settings: %w", err)
+	}
+
+	err = mapstructure.WeakDecode(write, &d.WriteSettings)
+	if err != nil {
+		return fmt.Errorf("failed to update write settings: %w", err)
+	}
+	return nil
+}
+
+type CreateTableOptions struct {
+	// View specifies whether the created table is a view.
+	View bool
+}
+
+type IncrementalStrategy string
+
+const (
+	IncrementalStrategyUnspecified IncrementalStrategy = ""
+	IncrementalStrategyAppend      IncrementalStrategy = "append"
+	IncrementalStrategyMerge       IncrementalStrategy = "merge"
+)
+
+type InsertTableOptions struct {
+	ByName    bool
+	Strategy  IncrementalStrategy
+	UniqueKey []string
+}
+
+// NewDB creates a new DB instance.
+// This can be a slow operation if the backup is large.
+// dbIdentifier is a unique identifier for the database reported in metrics.
+func NewDB(ctx context.Context, dbIdentifier string, opts *DBOptions) (DB, error) {
+	if dbIdentifier == "" {
+		return nil, fmt.Errorf("db identifier cannot be empty")
+	}
+	err := opts.ValidateSettings()
+	if err != nil {
+		return nil, err
+	}
+	db := &db{
+		dbIdentifier: dbIdentifier,
+		opts:         opts,
+		readPath:     filepath.Join(opts.LocalPath, "read"),
+		writePath:    filepath.Join(opts.LocalPath, "write"),
+		writeDirty:   true,
+		logger:       opts.Logger,
+	}
+	if opts.BackupProvider != nil {
+		db.backup = opts.BackupProvider.bucket
+	}
+	// if clean is true, remove the backup
+	if opts.Clean {
+		err = db.deleteBackup(ctx, "", "")
+		if err != nil {
+			return nil, fmt.Errorf("unable to clean backup: %w", err)
+		}
+	}
+
+	// create read and write paths
+	err = os.MkdirAll(db.readPath, fs.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create read path: %w", err)
+	}
+	err = os.MkdirAll(db.writePath, fs.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create write path: %w", err)
+	}
```

Comment:
Yes true. My idea was that keeping separate directories keeps the logic simpler in case of deletes and renames. 
Also we are never constrained by storage in duckDB so it should be fine to duplicate data. Let me revaluate `deletes` and `renames`.

Reply:
This will also run on local where disk storage is more sparse. If someone pulls 20 GB of data, taking of 40 GB of local disk space will be annoying. The cost of copying so many files back and forth is also likely to add up over time. So I think it would be best to just have one local cache of the remote data at a time.